swarm/network/stream/snapshot_retrieval_test.go

```diff
@@ -0,0 +1,378 @@
+			//If testing only history, we upload the chunk(s) first
+			conf.chunks, err = uploadFileToSingleNodeStore(uploadNode.ID(), chunkCount)
+			if err != nil {
+				return err
+			}
+		}
+
+		//variables needed to wait for all subscriptions established before uploading
+		errc := make(chan error)
+
+		//now setup and start event watching in order to know when we can upload
+		ctx, watchCancel := context.WithTimeout(context.Background(), MAX_TIMEOUT*time.Second)
+		defer watchCancel()
+
+		log.Info("Setting up stream subscription")
+		// each node Subscribes to each other's swarmChunkServerStreamName
+		for j, id := range ids {
+			log.Trace(fmt.Sprintf("Subscribe to subscription events: %d", j))
+			client, err := net.GetNode(id).Client()
+			if err != nil {
+				return err
+			}
+			//watch for peers disconnecting
+			err = streamTesting.WatchDisconnections(id, client, disconnectC, quitC)
+			if err != nil {
+				return err
+			}
+
+			watchSubscriptionEvents(ctx, id, client, errc)
+		}
+
+		for j, id := range ids {
+			log.Trace(fmt.Sprintf("Start syncing and stream subscriptions: %d", j))
+			client, err := net.GetNode(id).Client()
+			if err != nil {
+				return err
+			}
+			//start syncing!
+			var cnt int
+			err = client.CallContext(ctx, &cnt, "stream_startSyncing")
+			if err != nil {
+				return err
+			}
+			subscriptionCount += cnt
+			for snid := range registries[id].peers {
+				subscriptionCount++
+				err = client.CallContext(ctx, nil, "stream_subscribeStream", snid, NewStream(swarmChunkServerStreamName, "", false), nil, Top)
+				if err != nil {
+					return err
+				}
+			}
+		}
+
+		//now wait until the number of expected subscriptions has been finished
+		for err := range errc {
+			if err != nil {
+				return err
+			}
+			subscriptionCount--
+			if subscriptionCount == 0 {
+				break
+			}
+		}
+
+		log.Info("Stream subscriptions successfully requested, action terminated")
+
+		if live {
+			//now upload the chunks to the selected random single node
+			chnks, err := uploadFileToSingleNodeStore(uploadNode.ID(), chunkCount)
+			if err != nil {
+				return err
+			}
+			conf.chunks = append(conf.chunks, chnks...)
+		}
+
+		return nil
+	}
+
+	chunkSize := storage.DefaultChunkSize
+
+	//check defines what will be checked during the test
+	check := func(ctx context.Context, id discover.NodeID) (bool, error) {
+
+		if id == uploadNode.ID() {
+			return true, nil
+		}
+
+		select {
+		case <-ctx.Done():
+			return false, ctx.Err()
+		case e := <-disconnectC:
+			log.Error(e.Error())
+			return false, fmt.Errorf("Disconnect event detected, network unhealthy")
+		default:
+		}
+		log.Trace(fmt.Sprintf("Checking node: %s", id))
+		//if there are more than one chunk, test only succeeds if all expected chunks are found
+		allSuccess := true
+
+		dpa := registries[id].dpa
+		for _, chnk := range conf.chunks {
+			reader := dpa.Retrieve(chnk)
+			if s, err := reader.Size(nil); err != nil || s != chunkSize {
```

Comment:
chunkSize is not needed and just use a hash of a file of potentially many chunks.
see janos PR

Reply:
this particular test only uploads single random chunks. will add additional test which will upload file(s) to different nodes